File: spritebench_go/sprite.go
package main

import (
	"math"

	"graphics.gd/classdb/Node"
	"graphics.gd/classdb/SceneTree"
	"graphics.gd/classdb/Sprite2D"
	"graphics.gd/classdb/Texture2D"
	"graphics.gd/variant/Angle"
	"graphics.gd/variant/Float"
	"graphics.gd/variant/Vector2"
)

type Sprite struct {
	Sprite2D.Extension[Sprite]

	angle      Angle.Radians
	speed      Float.X
	pos        Vector2.XY
	size2      Vector2.XY
	windowSize Vector2.XY
}

func NewSprite(texture Texture2D.Instance, fromNode Node.Instance) *Sprite {
	s := &Sprite{}
	s.angle = Angle.Radians(Float.RandomBetween(0, math.Pi*2))
	s.speed = Float.RandomBetween(100, 600)
	windowSize := SceneTree.Get(fromNode).Root().AsWindow().Size()
	s.pos = Vector2.New(windowSize.X/2, windowSize.Y/2)
	s.AsSprite2D().SetTexture(texture)
	s.AsNode2D().SetPosition(s.pos)
	s.size2 = Vector2.MulX(s.AsSprite2D().Texture().GetSize(), 0.5)
	s.windowSize = Vector2.New(windowSize.X, windowSize.Y)
	return s
}

func (s *Sprite) Process(delta Float.X) {
	s.pos = Vector2.Add(s.pos, Vector2.XY{
		X: Angle.Cos(s.angle) * s.speed * delta,
		Y: Angle.Sin(s.angle) * s.speed * delta,
	})
	s.AsNode2D().SetPosition(s.pos)
	if s.pos.X < s.size2.X || s.pos.X > Float.X(s.windowSize.X)-s.size2.X {
		s.angle = Angle.Pi - s.angle
	}
	if s.pos.Y < s.size2.Y || s.pos.Y > Float.X(s.windowSize.Y)-s.size2.Y {
		s.angle = -s.angle
	}
}

File: spritebench_rust/rust/src/sprite.rs
use std::f32::consts::PI;
use std::f32;

use godot::classes::{ISprite2D, Sprite2D, Texture2D};
use godot::prelude::*;

#[derive(GodotClass)]
#[class(no_init, base=Sprite2D)]
pub struct BenchSprite {
    angle: f32,
    speed: f32,
    pos: Vector2,
    window_size: Vector2i,
    size_2: Vector2,

    base: Base<Sprite2D>,
}

#[godot_api]
impl BenchSprite {
    #[func]
    pub fn create_sprite(window_size: Vector2i, texture: Gd<Texture2D>) -> Gd<Self> {
        let mut sprite = Gd::from_init_fn(|base| Self {
            angle: rand::random_range(0.0..PI * 2.0) as f32,
            speed: rand::random_range(100.0..600.0) as f32,
            pos: Vector2 {
                x: window_size.x as f32 / 2.0,
                y: window_size.y as f32 / 2.0,
            },
            window_size,
            size_2: texture.get_size() / 2.0,
            base,
        });

        sprite.set_texture(&texture);

        sprite
    }
}

#[godot_api]
impl ISprite2D for BenchSprite {
    fn process(&mut self, delta: f64) {
        self.pos += Vector2 {
            x: f32::cos(self.angle),
            y: f32::sin(self.angle),
        } * self.speed * (delta as f32);

        let pos = self.pos;
        self.base_mut().set_position(pos);

        if self.pos.x < self.size_2.x || self.pos.x > self.window_size.x as f32 - self.size_2.x {
            self.angle = PI - self.angle;
        }
        if self.pos.y < self.size_2.y || self.pos.y > self.window_size.y as f32 - self.size_2.y {
            self.angle = -self.angle;
        }
    }
}
